src/observers.ts

```diff
@@ -0,0 +1,251 @@
+import {
+  DEFAULT_EPOCH_BLOCK_LENGTH,
+  DEFAULT_NUM_SAMPLED_BLOCKS,
+  DEFAULT_SAMPLED_BLOCKS_OFFSET,
+  GATEWAY_LEAVE_LENGTH,
+  MAX_TENURE_WEIGHT,
+  NUM_OBSERVERS_PER_EPOCH,
+  TENURE_WEIGHT_TOTAL_BLOCK_COUNT,
+} from './constants';
+import {
+  BlockHeight,
+  DeepReadonly,
+  Gateway,
+  Gateways,
+  RewardDistributions,
+  WeightedObserver,
+} from './types';
+
+export function getEpochBoundariesForHeight({
+  currentBlockHeight,
+  epochBlockLength = new BlockHeight(DEFAULT_EPOCH_BLOCK_LENGTH),
+  epochZeroBlockHeight,
+}: {
+  currentBlockHeight: BlockHeight;
+  epochBlockLength?: BlockHeight;
+  epochZeroBlockHeight: BlockHeight;
+}): {
+  epochStartHeight: BlockHeight;
+  epochEndHeight: BlockHeight;
+} {
+  const epochIndexForCurrentBlockHeight = Math.floor(
+    currentBlockHeight.valueOf() / epochBlockLength.valueOf(),
+  );
+  const epochStartHeight =
+    epochZeroBlockHeight.valueOf() +
+    epochBlockLength.valueOf() * epochIndexForCurrentBlockHeight;
+  return {
+    epochStartHeight: new BlockHeight(epochStartHeight),
+    epochEndHeight: new BlockHeight(
+      epochStartHeight + epochBlockLength.valueOf() - 1,
+    ),
+  };
+}
+
+// TODO: can we confidently us buffers here in non-node environments?
+export async function getEntropyHashForEpoch({
+  epochStartHeight,
+}: {
+  epochStartHeight: BlockHeight;
+}): Promise<Buffer> {
+  // used as we don't have access to Hash object in smartweave executions, so we concat our buffer and hash it at the end
+  let bufferHash: Buffer = Buffer.from('');
```

Comment:
we need to look at warp providing `node-polyfills` for browser environments. Otherwise `Buffer` would blow up on non `node` based environments